ChiModules/LinearBoltzmannSolver/Acceleration/diffusion_mip.h

```diff
@@ -0,0 +1,153 @@
+enum class BCType
+{
+  DIRICHLET = 1,
+  ROBIN = 2
+};
+
+/**Simple data structure to specify boundary conditions. Its stores the
+ * BC-type in `type` and an array of 3 values in `values`. For a
+ * Dirichlet-BC only `values[0]` is used to specify the value of the BC.
+ * For a robin boundary condition we use all 3 values in the form
+\f[
+a\phi + b \mathbf{n} \frac{\partial \phi}{\partial \mathbf{x}} = f
+\f]
+where \f$ a \f$, \f$ b \f$ and \f$ f \f$ map to `values[0]`, `values[1]` and
+`values[2]`, respectively.
+*/
+struct BoundaryCondition
+{
+  BCType type = BCType::DIRICHLET;
+  std::array<double,3> values = {0,0,0};
+};
+
+/**Since the solver below double for both WGDSA and TGDSA we define this
+ * simplified data structure to hold multigroup diffusion coefficients and
+ * removal cross sections only for the relevant groups. E.g., for WGDSA it
+ * will only hold the cross sections for the groupset, and for TGDSA there will
+ * only be one group (All groups collapsed into 1).*/
+struct Multigroup_D_and_sigR
+{
+  std::vector<double> Dg;
+  std::vector<double> sigR;
+};
+
+/**Generalized diffusion solver for both WGDSA and TGDSA based on the MIP-method
+ * of Bruno Turcksin and Jean Ragusa.*/
+class DiffusionMIPSolver
+{
+protected:
+  typedef std::map<int,Multigroup_D_and_sigR> MapMatID2XS;
+protected:
+  const std::string m_text_name;
+  const chi_mesh::MeshContinuum& m_grid;
+  const chi_math::SpatialDiscretization& m_sdm;
+  const chi_math::UnknownManager m_uk_man;
+
+  const std::vector<BoundaryCondition> m_bcs;
+
+  const MapMatID2XS m_map_mat_id_2_xs;
+
+  const std::vector<UnitCellMatrices>& m_unit_cell_matrices;
+
+  const int64_t m_num_local_dofs;
+  const int64_t m_num_global_dofs;
+
+  Mat m_A = nullptr;
+  Vec m_rhs = nullptr;
+  KSP m_ksp = nullptr;
+
+public:
+  struct Options
+  {
+    double      residual_tolerance = 1.0e-4; ///< Residual tol. relative to rhs
+    int         max_iters = 100;             ///< Maximum iterations
+    bool        verbose = false;             ///< Verbosity flag
+    bool        perform_symmetry_check = false; ///< For debugging only (very expensive)
+    std::string source_lua_function;       ///<for mms
+    std::string ref_solution_lua_function; ///<for mms
+    std::string additional_options_string;
+  }options;
+public:
+  const chi_math::UnknownManager& UnknownStructure() const
+  {
+    return m_uk_man;
+  }
+  //00
+  DiffusionMIPSolver(std::string text_name,
+                     const chi_mesh::MeshContinuum& grid,
+                     const chi_math::SpatialDiscretization& sdm,
+                     const chi_math::UnknownManager& uk_man,
+                     std::vector<BoundaryCondition> bcs,
+                     MapMatID2XS map_mat_id_2_xs,
+                     const std::vector<UnitCellMatrices>& unit_cell_matrices,
+                     bool verbose=false);
+  //00a
+  void Initialize();
+
+  //01a
+  void AssembleAand_b_wQpoints(const std::vector<double>& q_vector);
```

Comment:
for the names, should there be a _ before A to be cosnistent with other names?

Reply:
We actually allow this. This is what is written in the coding conventions section of the manual.

FunctionNames have no underscores and each word starts with a capital letter. Underscores can be allowed in some circumstances.